fjord/base/l10n.py

```diff
@@ -42,54 +41,37 @@ def __init__(self, environment):
+
     def _parse_block(self, parser, allow_pluralize):
-        parse_block = InternationalizationExtension._parse_block
-        ref, buffer = parse_block(self, parser, allow_pluralize)
-        return ref, collapse_whitespace(buffer)
+        """parses the trans block
 
+        This overrides trans parsing to look for a "trimmed" variable. If
+        it's there, then this works just like Django's blocktrans and it'll
+        collapse the whitespace in the trans tag.
 
-def tweak_message(message):
-    """We piggyback on jinja2's babel_extract() (really, Babel's extract_*
-    functions) but they don't support some things we need so this function will
-    tweak the message.  Specifically:
+        For example, this is not collapsed::
 
-    1. We collapse whitespace in the msgid. Jinja2 will only strip
-       whitespace from the ends of a string so linebreaks show up in
-       your .po files still.
+            {% trans %} ... {% endtrans %}
 
-    2. Babel doesn't support context (msgctxt). We hack that in ourselves
-       here.
+        This is collapsed::
 
-    """
-    if isinstance(message, basestring):
-        message = collapse_whitespace(message)
-    elif isinstance(message, tuple):
-        # A tuple of 2 has context, 3 is plural, 4 is plural with context
-        if len(message) == 2:
-            message = add_context(message[1], message[0])
-        elif len(message) == 3:
-            if all(isinstance(x, basestring) for x in message[:2]):
-                singular, plural, num = message
-                message = (collapse_whitespace(singular),
-                           collapse_whitespace(plural),
-                           num)
-        elif len(message) == 4:
-            singular, plural, num, ctxt = message
-            message = (add_context(ctxt, collapse_whitespace(singular)),
-                       add_context(ctxt, collapse_whitespace(plural)),
-                       num)
-    return message
-
-
-def extract_python(fileobj, keywords, comment_tags, options):
-    msgs = list(babel_extract_python(fileobj, keywords, comment_tags, options))
-    for lineno, funcname, message, comments in msgs:
-        message = tweak_message(message)
-        yield lineno, funcname, message, comments
-
-
-def extract_template(fileobj, keywords, comment_tags, options):
-    msgs = list(ext.babel_extract(fileobj, keywords, comment_tags, options))
-    for lineno, funcname, message, comments in msgs:
-        message = tweak_message(message)
-        yield lineno, funcname, message, comments
+            {% trans trimmed %} ... {% endtrans %}
+
+        And this is also collapsed::
+
+            {% trans foo=bar, trimmed %} ... {% endtrans %}
+
+        """
+        parse_block = InternationalizationExtension._parse_block
+        ref, buffer = parse_block(self, parser, allow_pluralize)
+
+        if self._yolo_is_trimmed():
+            buffer = collapse_whitespace(buffer)
+        return ref, buffer
```

Comment:
This new `_parse_block` only collapses whitespace if there's a "trimmed" in the thingy. For Django blocktrans tags, that "thingy" is a flag indicating that the whitespace should be collapsed. Here, it's a variable and we're (ab)using inspect and stackframes to see whether it's there and collapsing accordlingly.

The nice thing this affords us is this:

1. You can use tower's extract (sans all the "collapse all the whitespace!") command. It uses babel which compiles the Jinja2 template to AST and then walks the AST for trans nodes. Since we're collapsing the whitespace in the block, it picks up the text and that's the msgid. This works when rendering the template, too. Everything is rainbows.
2. You can use django-jinja's makemessages which overrides django's makemessages. It's looking for things that look like django's stuff, so it sees the "trimmed" and it collapses the whitespace. That's "good enough".

Further, this lets us specify which blocks should be trimmed. I'm not sure why you would want to do otherwise. I kind of think this is a stupid default. But this is closer to the Django behavior.